### ISSUE TITLE
Coins sent to wrong address

### ISSUE THREAD

**Issue by ASAPSegfault** [bearby-wallet/bearby-extension#62]
When using bearby configured for a local node : 
![image](https://github.com/bearby-wallet/bearby-extension/assets/45430752/e83297a4-84d8-4451-97de-a5d67531bf6e)

And sending some coins to a smart contract, everything looks fine the frontend : 
![image](https://github.com/bearby-wallet/bearby-extension/assets/45430752/ad0724d1-c0b6-4645-a1f0-9018490f63e8)

But in reality when I check in the node, here's what happens :
![image](https://github.com/bearby-wallet/bearby-extension/assets/45430752/10b5ed16-6aa1-425c-8bed-d431dcf27300)

In other words, the recipient address `AS12VJfZVX8fZN2H7DTThnKGphcEMsbc8Cb1PUddKKKppnn7mfRPo `gets replaced by this address : 
`AU12VJfZVX8fZN2H7DTThnKGphcEMsbc8Cb1PUddKKKppnn7mfRPo`. It's the exact same but the S get replace by a U.

**Comment by hicaru** [comment]
Thanks so much for your issue, it very important  i will fix it in new release!

**Comment by hicaru** [comment]
## I will forward this issue to the Massa core team.

I have thoroughly examined the issue, and it's not a problem with the Bearby Wallet; the issue lies in the nodes. They incorrectly process the address because, when sending, they do not know which prefix was used. For the nodes, whether it's AU or AS, the address appears the same.

i use massa-web3 for testing it.
```js
import { CHAIN_ID, ClientFactory, ProviderType, WalletClient, fromMAS } from "@massalabs/massa-web3";

(async function() {
  const sender = await WalletClient.getAccountFromSecretKey('your privateKey');
  const web3Client = await ClientFactory.createCustomClient(
    [
      { url: 'https://buildnet.massa.net/api/v2', type: ProviderType.PUBLIC }
    ],
    CHAIN_ID.BuildNet,
    true,
    sender
  );

  await web3Client.wallet().addAccountsToWallet([sender]);
  const accountBalance = await web3Client
    .wallet()
    .getAccountBalance(sender.address);

  console.log(accountBalance);

  const txId = await web3Client.wallet().sendTransaction({
    amount: fromMAS(1),
    fee: 0n,
    recipientAddress: 'AS12VJfZVX8fZN2H7DTThnKGphcEMsbc8Cb1PUddKKKppnn7mfRPo',
  });

  console.log(txId);
}());

```
i got same result 
<img width="850" alt="image" src="https://github.com/bearby-wallet/bearby-extension/assets/21293723/6b026847-749b-440e-ba9b-f10ff442cee2">

**Comment by damip** [comment]
Thanks for reporting. We are looking into it

**Comment by Leo-Besancon** [comment]
Hi @ASAPSegfault and @hicaru !
Thanks for pointing this out. What I found is that this does not seem to be an issue with the node.
Here are a few things for context:
- Using a SendTransaction operation with a smart contract address as recipient should lead to an error. We want to prevent users from sending tokens to smart contracts that do not explicitly handles it.
- When I use the client to send a transaction to a smart contract address, the operation has the correct recipient and the execution fails.
Here is the event trace:
```
Data: {"massa_execution_error":"Runtime error: runtime error when executing operation O15EyxMmN4avAAJqRbPHS1d6UM77hCWZzuYavuafWV4HDPv96fx: `Transaction` error: transfer of 1 coins from AU1CCxYnttgXg4dWn1ADh4z4scXX7JreYPE9EEkM2FMroREQAMfA to AS1RzQ5Jacoaz3AcT8PxCBN9uit8rXJjNeNxaqDagT9zPnU5EQnX failed: Runtime error: crediting SC address AS1RzQ5Jacoaz3AcT8PxCBN9uit8rXJjNeNxaqDagT9zPnU5EQnX is not allowed without write access to it"}
```
- Massa-web3 does have the same issue, so that's why your tests show the same results between bearby and massa-web3! https://github.com/massalabs/massa-web3/issues/533
- In the following code, when creating the bytes of the recipient address, you seem to assume it is a user address: https://github.com/bearby-wallet/bearby-extension/blob/master/core/background/provider/transaction.ts#L44
What you should do instead is parse the prefix of the address string. If it's AU, serialize the bytes as a user address. If it's AS, serialize the bytes as a contract address like here: https://github.com/bearby-wallet/bearby-extension/blob/master/core/background/provider/transaction.ts#L296

**Comment by hicaru** [comment]
> Hi @ASAPSegfault and @hicaru ! Thanks for pointing this out. What I found is that this does not seem to be an issue with the node. Here are a few things for context:
> 
>     * Using a SendTransaction operation with a smart contract address as recipient should lead to an error. We want to prevent users from sending tokens to smart contracts that do not explicitly handles it.
> 
>     * When I use the client to send a transaction to a smart contract address, the operation has the correct recipient and the execution fails.
>       Here is the event trace:
> 
> 
> ```
> Data: {"massa_execution_error":"Runtime error: runtime error when executing operation O15EyxMmN4avAAJqRbPHS1d6UM77hCWZzuYavuafWV4HDPv96fx: `Transaction` error: transfer of 1 coins from AU1CCxYnttgXg4dWn1ADh4z4scXX7JreYPE9EEkM2FMroREQAMfA to AS1RzQ5Jacoaz3AcT8PxCBN9uit8rXJjNeNxaqDagT9zPnU5EQnX failed: Runtime error: crediting SC address AS1RzQ5Jacoaz3AcT8PxCBN9uit8rXJjNeNxaqDagT9zPnU5EQnX is not allowed without write access to it"}
> ```
> 
>     * Massa-web3 does have the same issue, so that's why your tests show the same results between bearby and massa-web3! [SendTransaction operation recipient is assumed to be a UserAddress, but should fail if it's not massalabs/massa-web3#533](https://github.com/massalabs/massa-web3/issues/533)
> 
> 
> In the following code, when creating the bytes of the recipient address, you seem to assume it is a user address: https://github.com/bearby-wallet/bearby-extension/blob/master/core/background/provider/transaction.ts#L44 What you should do instead is parse the prefix of the address string. If it's AU, serialize the bytes as a user address. If it's AS, serialize the bytes as a contract address like here: https://github.com/bearby-wallet/bearby-extension/blob/master/core/background/provider/transaction.ts#L296

The underlying issue seems to originate from the node itself. Despite implementing rigorous controls, users are still finding ways to circumvent these measures. This challenge spans across blockchains like Ethereum, Solana, and Polygon, where operations are reliant on specific methods within smart contracts.

When it comes to address serialization, the node appears indifferent to the choice of address I employ. This indicates a definitive bug within the protocol, especially considering that similar issues are not present in other blockchain networks.

**Comment by Leo-Besancon** [comment]
Hi!

> When it comes to address serialization, the node appears indifferent to the choice of address I employ. This indicates a definitive bug within the protocol, especially considering that similar issues are not present in other blockchain networks.

I'm not sure I understand this point. Can you detail exactly where you serialize an address and the node does not take the type provided correctly?

The addresses have two formats:
- Strings: starts with 'AU' for user addresses and 'AS for smart contract addresses
- Bytes: starts with 0 for user addresses and 1 for smart contract addresses (encoded as varints).

Then, it depends on what you need.
- If you send the string address to the node, it will decode itself the prefix (AU or AS) to know which type it is.
- If you need to serialize the address bytes (like here: https://github.com/bearby-wallet/bearby-extension/blob/8edf23b2457b8f6d40392fbe7b70ff68424f75c9/core/background/provider/transaction.ts#L44),
then you have to decode the AU or AS prefix yourself in order to correctly put 0 or 1 as the bytes prefix. Here, it seems that you use the user address bytes serialization even if the address starts with AS.

I would use a helper like that instead (I have not tested it):

I think you should implement a helper:
```js
async function convertStringAddressToBytes (addr_str) => {
  let addr_bytes = (await base58Decode(addr_str.slice(ADDRESS_PREFIX.length)));
  const type_prefix = addr_str.slice(0, ADDRESS_PREFIX.length);
  
  if (type_prefix === ADDRESS_PREFIX) {
	addr_bytes = Uint8Array.from([VERSION_NUMBER, ...addr_bytes]);
  } else if (type_prefix === CONTRACT_ADDRESS_PREFIX ) {
	addr_bytes = Uint8Array.from([CONTRACT_VERSION_NUMBER, ...addr_bytes]);
  } else {
	throw new Error("Invalid address type");
  }
  return addr_bytes;
}
```

Please let me know if something is not clear!
